src/metabase/models/params/custom_values.clj

```diff
@@ -0,0 +1,110 @@
+(ns metabase.models.params.custom-values
+  (:require
+    [clojure.string :as str]
+    [metabase.mbql.normalize :as mbql.normalize]
+    [metabase.models :refer [Card]]
+    [metabase.mbql.schema :as mbql.s]
+    [metabase.query-processor :as qp]
+    [metabase.search.util :as search]
+    [metabase.util.i18n :refer [tru]]
+    [metabase.util.schema :as su]
+    [schema.core :as s]
+    [toucan.db :as db]))
+
+;; -------------------------------------- source=card --------------------------------------
+
+(def ^{:dynamic true
+       :doc     "Maximum number of rows returned when running a card."}
+  *max-rows* 2000)
+
+(defn- field-ref->mbql-field
+  "Find field by field_ref.
+
+  If field_ref is a:
+  - field: returns itself
+  - aggregation: returns a field using name [:field aggregation-name nil]"
+  [field-ref result-metadata card-id]
+  (let [field (first (filter #(= (:field_ref %) field-ref) result-metadata))]
+    (when-not field
+      (throw (ex-info (tru "No matching field found")
+                      {:status-code 400
+                       :card-id     card-id
+                       :field-ref   field-ref})))
+    (case (first field-ref)
+      :aggregation
+      [:field (:name field) {:base-type ((some-fn :effective_type :base_type) field)}]
+
+      :field
+      (let [[_ttype identifier options] field-ref]
+        ;; TODO: the options for field-ref does contains the binning information
+        ;; maybe we could just need to select it from options?
+        (when (some? (:binning options))
+              (throw (ex-info (tru "Binning column not supported")
+                              {:status-code 400
+                               :card-id     card-id
+                               :field-ref   field-ref})))
+        [:field identifier (select-keys options mbql.s/field-options-for-identification)])
+
+      (throw (ex-info (tru "Invalid field-ref type. Must be a field or aggregation.")
+                      {:status-code 400
+                       :card-id     card-id
+                       :field-ref   field-ref})))))
```

Comment:
Throwing exceptions with `:status-code` couples this function with api endpoints. If code is in a `metabase.models` namespace I would have assumed it should be decoupled from the api. So it should either be impossible to reach this code path (in that case this exception doesn't need to be thrown), or it shouldn't live in the `metabase.models` namespace. The exception for binning seems fine as long as we clean it up later. btw I think this is an existing issue with `metabase.models.params.chain-filter` too.

Reply:
I hear what you're saying but I think we don't have strict rules on whether the exception for API  should only be placed in the API namespace.

If we want to do that then pretty much any call to `models` namespace will need a try-catch and proper exception naming so API namespace could selectively catch what is intentionally thrown.

FWIW a lot of models namespace throw status-code too: permissions, collection, card to name a field.
I'm saying this not to mean it's okay to do so, but unless we have a better error-handling mechanism I think it's okay to throw them in models ns for now.


On whether this code path is reached or not: yes, I didn't have a test for it but I have a test for the `no matching field found` case: https://github.com/metabase/metabase/blob/afb4eb38b7afd5f96dfcbd8a31556535c343c58c/test/metabase/api/dashboard_test.clj#L2116

Reply:
ah, you're right it's not reachable for that exception :+1. Will remove that throw.

On throw status-code from model ns. I think it's a general problem.
My idea of how we should throw this is we need to have custom exception so that the API can catches only thing it meant to catch( in this case no matching field found so it knows it should returns 400)

Otherwise we'll have to have API catch generic exception `clojure.lang.ExceptionInfo` which will catches all exceptions and just return a generic error. not to mention it doesn't know if it should be a 400 or 404?

to sum up, I agree throwing status-code in model ns is not ideal, but it's not the end of the world given how we're currently doing it across codebase.

Reply:
I'd still argue 400 errors should be caught at the API layer. In this case the binning exception is okay since that's meant to be temporary and it's not worth fixing in the short term.